source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+  if (!connect_finished_) {
+    per_connection_state_.detect_early_close_when_read_disabled_ = value;
+  }
+  for (auto& connection : connections_) {
+    connection->detectEarlyCloseWhenReadDisabled(value);
+  }
+}
+
+bool HappyEyeballsConnectionImpl::readEnabled() const {
+  if (!connect_finished_) {
+    return !post_connect_state_.read_disable_count_.has_value() ||
+           post_connect_state_.read_disable_count_ == 0;
+  }
+  return connections_[0]->readEnabled();
+}
+
+const SocketAddressProvider& HappyEyeballsConnectionImpl::addressProvider() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->addressProvider();
+}
+
+SocketAddressProviderSharedPtr HappyEyeballsConnectionImpl::addressProviderSharedPtr() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->addressProviderSharedPtr();
+}
+
+absl::optional<Connection::UnixDomainSocketPeerCredentials>
+HappyEyeballsConnectionImpl::unixSocketPeerCredentials() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->unixSocketPeerCredentials();
+}
+
+Ssl::ConnectionInfoConstSharedPtr HappyEyeballsConnectionImpl::ssl() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->ssl();
+}
+
+Connection::State HappyEyeballsConnectionImpl::state() const {
+  if (!connect_finished_) {
+    ASSERT(connections_[0]->state() == Connection::State::Open);
+  }
+  return connections_[0]->state();
+}
+
+bool HappyEyeballsConnectionImpl::connecting() const {
+  ASSERT(connect_finished_ || connections_[0]->connecting());
+  return connections_[0]->connecting();
+}
+
+void HappyEyeballsConnectionImpl::write(Buffer::Instance& data, bool end_stream) {
+  if (connect_finished_) {
+    connections_[0]->write(data, end_stream);
+    return;
+  }
+
+  post_connect_state_.write_buffer_ = dispatcher_.getWatermarkFactory().createBuffer(
+      []() -> void { ASSERT(false); }, [this]() -> void { this->onWriteBufferHighWatermark(); },
+      []() -> void { ASSERT(false); });
+  if (per_connection_state_.buffer_limits_.has_value()) {
+    post_connect_state_.write_buffer_.value()->setWatermarks(
+        per_connection_state_.buffer_limits_.value());
+  }
+  post_connect_state_.write_buffer_.value()->move(data);
+  post_connect_state_.end_stream_ = end_stream;
+}
+
+void HappyEyeballsConnectionImpl::setBufferLimits(uint32_t limit) {
+  if (!connect_finished_) {
+    ASSERT(!per_connection_state_.buffer_limits_.has_value());
+    per_connection_state_.buffer_limits_ = limit;
+    if (post_connect_state_.write_buffer_.has_value()) {
+      post_connect_state_.write_buffer_.value()->setWatermarks(
+          per_connection_state_.buffer_limits_.value());
+    }
+  }
+  for (auto& connection : connections_) {
+    connection->setBufferLimits(limit);
+  }
+}
+
+uint32_t HappyEyeballsConnectionImpl::bufferLimit() const {
+  if (!connect_finished_) {
```

Comment:
can always just return connection[0]'s no?

Reply:
Yeah, that seems reasonable! Done.